### PR TITLE
Precompile triton templates

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122643
* #122642
* #121999
* __->__ #121998

Before this PR we were not precompiling triton templates in parallel. Compilation would occur during benchmarking.

Triton benchmarking templates were emitted as :

```
@triton.jit
def triton_mm(arg_A, arg_B, out_ptr0):
```

In order to precompile we need to give the full kernel specification, as we do when we emit the template in the final output code generation.

```
@triton_heuristics.template(
    num_stages=3,
    num_warps=8,
    triton_meta={'signature': {0: '*fp32', 1: '*fp32', 2: '*fp32'}, 'device': 0, 'device_type': 'cuda', 'constants': {}, 'configs': [AttrsDescriptor(divisible_by_16=(0, 1, 2), equal_to_1=(), ids_of_folded_args=(), divisible_by_8=())]},
    inductor_meta={'kernel_name': 'Placeholder.DESCRIPTIVE_NAME', 'backend_hash': 'cdeecfeccd31ad7810f96b5752194b1c2406d0a81e39a6ca09c8ee150baae183'},
)
@triton.jit
def triton_mm(arg_A, arg_B, out_ptr0):
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by jansel — APPROVED** [review]

**Comment by shunting314** [comment]
Trying to understand more, why the full kernel specification is required if we want to precompile in parallel?

File: torch/_inductor/select_algorithm.py [hunk 1]
```diff
@@ -498,7 +498,7 @@ def generate(
         ), TritonTemplateKernel(
             kernel_name=kernel_name,
             output_node=fake_out,
-            use_jit=True,
+            use_jit=False,
```
Comment (shunting314) [inline torch/_inductor/select_algorithm.py:501]: After this, is there any remaining cases that set `use_jit` to True? If not, we can probably remove this argument

**Comment by eellison** [comment]
@shunting314 in order to compile with JIT you need actually invoke it. There is a "warmup_only" arg in triton we could use, however we'd still need to provide real tensor inputs. 

In the case that you have arbitrarily many mms you want to compile in parallel, then you are potentially allocating the arguments for all of them at once which is not feasible for memory.

We already generate the full argument specification when we codegen anyway, so it's better to reuse this then the alternative.

**Comment by shunting314** [comment]
I see. You actually want the precompile ability of triton_heuristics.template

**Pull request closed without merging**